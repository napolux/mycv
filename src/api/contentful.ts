import * as Contentful from 'contentful';

export interface IContentfulApiClient {
  client: Contentful.ContentfulClientApi;
  getCv(): Promise<any>;
  mapResponseToPositions(positionsFromResponse: any): Array<ICvPositions>
}

export interface ICvFields {
  name: string;
  surname: string;
  dob: Date;
  phone: string;
  email: string;
  positions: ICvPositions[];
}

export interface ICvPositions {
  company: string;
  role: string;
  description: string;
  start: Date;
  end: Date;
}

export class ContentfulAPIClient implements IContentfulApiClient {

  client: Contentful.ContentfulClientApi;

  constructor(space: string, accessToken: string) {
    this.client = Contentful.createClient({
      space: space,
      accessToken: accessToken
    })
  }

  // We'll get the CV from Contentful
  getCv = async () => {
    try {
      const cv = await this.client.getEntry('2DMIcTZmJvEUe2FFpEQZaT');
      return cv;  
    } catch (error) {
      return {};
    }
  }

  // Utility to map response job positions to output job positions
  mapResponseToPositions(positionsFromResponse: any): Array<ICvPositions> {
    
    let positions: Array<ICvPositions> = [];
    
    positionsFromResponse.forEach((p: any) => {
      positions.push(
        {
          start: p.fields.start,
          end: p.fields.end,
          company: p.fields.company,
          role: p.fields.role,
          description: p.fields.description
        }
      )
    });

    return positions;
  } 
}