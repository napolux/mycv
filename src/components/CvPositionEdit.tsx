import React from 'react';
import { ICvPositions } from '../api/contentful';
import { Form, Divider, Container, Modal, Button } from 'semantic-ui-react';

interface ICvPositionProps {
  cvPositionValue: ICvPositions[];
}

interface ICvPositionState {
  modalOpen: boolean;
  hidden: boolean
}

export default class CvPositionEdit extends React.Component<ICvPositionProps, ICvPositionState> {

  constructor(props: ICvPositionProps) {
    super(props);
    this.state = { modalOpen: false, hidden: false };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.closeModal = this.closeModal.bind(this);
  }

  private closeModal() {
    this.setState({ modalOpen: false });
  }

  private handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    this.setState({
      modalOpen: true
    });
  }

  render() {

    // for readability
    let p = this.props.cvPositionValue;

    // job positions to be rendered
    let positions = [];

    for (let i = 0; i < p.length; i++) {
      positions.push(
        <Container key={'cvposition_' + i}>
          <div>
            <Form method="post" onSubmit={this.handleSubmit}>
              <Form.Field>
                <label htmlFor="company">Company:</label>
                <input id="company" name="company" defaultValue={p[i].company} />
              </Form.Field>
              <Form.Field>
                <label htmlFor="role">Role:</label>
                <input id="role" name="role" defaultValue={p[i].role} />
              </Form.Field>
              <Form.Field>
                <label htmlFor="start">Start:</label>
                <input id="start" name="start" defaultValue={p[i].start.toString()} />
              </Form.Field>
              <Form.Field>
                <label htmlFor="end">End:</label>
                <input id="end" name="end" defaultValue={p[i].end.toString()} />
              </Form.Field>
              <Form.Field>
                <label htmlFor="description">Description:</label>
                <input id="description" name="description" defaultValue={p[i].description} />
              </Form.Field>
              <Button color="green" size="mini">Save</Button>
            </Form>
          </div>
          <Divider />
        </Container>
      );
    }
    return (
      <Form method="post" onSubmit={this.handleSubmit}>
        {positions}
        <Modal size="mini" open={this.state.modalOpen} onClose={this.closeModal}>
          <Modal.Header>Content saved</Modal.Header>
          <Modal.Content>
            <p>Operation was successful</p>
          </Modal.Content>
        </Modal>
      </Form>
    );
  }
}