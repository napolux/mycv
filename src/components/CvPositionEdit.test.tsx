import React from 'react';
import CvPositionEdit from './CvPositionEdit';
import renderer from 'react-test-renderer';
import { ICvPositions } from '../api/contentful';

it('renders correctly', () => {

  const position: ICvPositions[] = [{
    company: 'mock company',
    role: 'mock role',
    description: 'string',
    start: new Date(2018, 11, 24, 10, 33, 30, 0),
    end: new Date(2018, 11, 24, 10, 33, 30, 0),
  }]

  const tree = renderer
    .create(<CvPositionEdit cvPositionValue={position} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});