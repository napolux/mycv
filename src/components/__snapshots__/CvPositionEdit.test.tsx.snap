// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="ui form"
  method="post"
  onSubmit={[Function]}
>
  <div
    className="ui container"
  >
    <div>
      <form
        className="ui form"
        method="post"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <label
            htmlFor="company"
          >
            Company:
          </label>
          <input
            defaultValue="mock company"
            id="company"
            name="company"
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="role"
          >
            Role:
          </label>
          <input
            defaultValue="mock role"
            id="role"
            name="role"
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="start"
          >
            Start:
          </label>
          <input
            defaultValue="Mon Dec 24 2018 10:33:30 GMT+0000 (Coordinated Universal Time)"
            id="start"
            name="start"
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="end"
          >
            End:
          </label>
          <input
            defaultValue="Mon Dec 24 2018 10:33:30 GMT+0000 (Coordinated Universal Time)"
            id="end"
            name="end"
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="description"
          >
            Description:
          </label>
          <input
            defaultValue="string"
            id="description"
            name="description"
          />
        </div>
        <button
          className="ui green mini button"
          onClick={[Function]}
        >
          Save
        </button>
      </form>
    </div>
    <div
      className="ui divider"
    />
  </div>
</form>
`;
